File: uoj_455_periodic-strings.cpp
#include <iostream>
#include <vector>
#include <string>

int main() {
    int cases;
    std::cin >> cases;
    std::cin.ignore();
    
    for (int caseNum = 0; caseNum < cases; caseNum++) {
        std::string myString;
        std::cin.ignore(); // Consume the newline after cases
        std::getline(std::cin, myString);
        
        std::vector<int> startingPoints;
        startingPoints.push_back(0);
        
	// std::cout << "Case: " << caseNum << "\t" + myString << std::endl;
        for (int index = 1; index < myString.length(); index++) 
            if (myString[0] == myString[index]) 
                startingPoints.push_back(index);
        
        bool periodFound = false;
	// For every other character in the string.
        for (int index = 1; index < myString.length(); index++)
	  {
            char currentChar = myString[index];
            if (periodFound) 
                break;
	    // std::cout << "Starting with index " << index << std::endl;

	    // Clever hack to iterate startingPoints
            for (int i = 0; i < startingPoints.size(); )
	      {
		// std::cout << "startingPoint " << startingPoints[i] << std::endl;
		if (startingPoints[i]+index < myString.length())
		  {
		    // If this is not a valid starting point, delete
		    // it, otherwise, move to the next (i++)
                    if (myString[startingPoints[i] + index]
			!= currentChar) 
                        startingPoints.erase(startingPoints.begin() + i);
		    else 
		      i++;
		  }
		else
		  {
                    if (index * startingPoints.size()
			== myString.length() ||
			startingPoints.size() == 1) 
		      {
			periodFound = true;
			break;
                       } 
		    else
		      startingPoints.erase(startingPoints.begin() + i);
		  }
	      }
        }
        
        if (caseNum > 0) {
            std::cout << std::endl;
        }


        if (startingPoints.size() > 1) {
            std::cout << startingPoints[1] << std::endl;
        } else {
            std::cout << myString.length() << std::endl;
        }
    }
    
    return 0;
}

File: uoj_455_periodic-strings.py
import string
# import map

cases = int(input())

for case in range(cases):
    input() # blank line
    myString = input()

    # I want to find an algorithm that works with indices. Can this be
    # thought of as a divide and conquer problem? Not really, but
    ## Suppose that we have a string such as "abvsasdfabvsasdf":
    ## - If I know that it is periodic, I only 

    # 1. Iterate through the string, storing the index of each
    # occurence of a number in an ordered set map

    # 2. For every character
    ## From the list of current periods, eliminate any that don't fit

    # 3. If period substrings are left, keep the smallest one. Else
    # period is the length of the starting string
    startingPoints = [0]

    for index in range(1,len(myString)):
        if myString[0] == myString[index]:
            startingPoints.append(index)

    periodFound = False
    for index in range(1,len(myString)):
        currentChar = myString[index]
        if periodFound:
            break
        for startingPoint in startingPoints:
            try:
                if myString[startingPoint+index] != currentChar:
                    startingPoints.remove(startingPoint)
            except Exception:
                # print(f"Error: {case+1}")
                # print(f"Error: Index{index}, Length: {len(startingPoints)}")
                # print(f"Error: Input length: {len(myString)}")
                if (index)*len(startingPoints) == len(myString):
                    # print(f"\tFound smallest period")
                    periodFound = True
                else:
                    startingPoints.remove(startingPoint)
        
    # At this point, all we need is the second starting point
    if case:
        print()
    if len(startingPoints) > 1:
        print(f"{startingPoints[1]}")
    else:
        print(f"{len(myString)}")
